TFEngine.py

```diff
@@ -1344,48 +1345,84 @@ def _do_save(self):
+        "Only %s are allowed in function eval_model as output_per_seq_format, but got: %s " % (allowed_outputs, output_per_seq_format)
+
+      # always fetch seq_tag to map loss values to the corresponding line
+      extra_fetches = {"seq_tags": self.network.get_seq_tags()}
+
       from TFUtil import identity
       losses_dict, _, _ = self.network.get_losses_initialized(reduce_func=identity, with_total=False)
-      for loss_name, loss_holder in losses_dict.items():
-        loss_holder.loss.layer = None  # enforce reinit
-        loss_holder.loss.use_flatten_frames = False  # we need that such that we get (B*T,...) unreduced values
-        loss_value = loss_holder.get_normalized_loss_value_per_seq()
-        error_value = loss_holder.get_normalized_error_value_per_seq()
-        if loss_value is not None:
-          extra_fetches["loss:score:%s" % loss_name] = loss_value
-        if error_value is not None:
-          extra_fetches["loss:error:%s" % loss_name] = error_value
-        if loss_holder.loss.output.have_time_axis():
-          seq_lens = loss_holder.loss.output.get_sequence_lengths()
-          extra_fetches["loss:seq_lens:%s" % loss_name] = seq_lens
-        extra_fetches["seq_tags"] = self.network.get_seq_tags()
-      assert extra_fetches
-
+      assert loss_name in losses_dict, \
+        "Unknown loss defined. Got \"%s\". Possible losses are %s" % (loss_name, losses_dict.keys())
+
+      #for loss_name, loss_holder in losses_dict.items():
+      loss_holder = losses_dict[loss_name]
+      loss_holder.loss.layer = None                  # enforce reinit, otherwise the new value of 'loss_holder.loss.use_flatten_frames' will be ignored
+      loss_holder.loss.use_flatten_frames = False    # we need that such that we get (B*T,...) unreduced values
+
+      # we need sequence lengths for positional fetches
+      has_positional_fetch = ("pos_score" in requested_outputs) or ("pos_error" in requested_outputs)
+
+      if "seq_len" in requested_outputs or has_positional_fetch:
+        extra_fetches["seq_len"] = loss_holder.loss.output.get_sequence_lengths()
+      if "score" in requested_outputs:
+        extra_fetches["score"] = loss_holder.get_normalized_loss_value_per_seq()
+      if "error" in requested_outputs:
+        extra_fetches["error"] = loss_holder.get_normalized_error_value_per_seq()
+      if "pos_score" in requested_outputs:
+        extra_fetches["pos_score"] = loss_holder.get_normalized_loss_value_per_seq(per_pos=True)
+      if "pos_error" in requested_outputs:
+        extra_fetches["pos_error"] = loss_holder.get_normalized_error_value_per_seq(per_pos=True)
+
+    results_per_seq = {}  # seq_tag -> dict[str,float]. Results of fetches will be written in this dict
+
+    # function to save the return values of each callback to the dict `results_per_seq`
     def extra_fetches_callback(seq_tags, **extra_fetches_out):
       """
-      :param list[str] seq_tags:
+      :param list[str] seq_tags
```

Comment:
This syntax is wrong. Why did you remove the `:`?
(Btw, curious, does PyCharm not print a warning/hint about this? E.g. like that `seq_tags` is not documented? But what you should notice in any case is that now PyCharm will not know the type of `seq_tags` anymore, which is something which you should always check, and always try to avoid if possible.)

Reply:
Ok. But always check you commits also, line by line, if your change is really meaningful. Some change like this should be seen this way.